hydra-node/src/Hydra/Chain/Direct.hs

```diff
@@ -58,56 +158,280 @@ import Ouroboros.Network.Protocol.LocalTxSubmission.Server (
+  -- TODO: Provide tracers for these.
+  tracers :: NetworkConnectTracers LocalAddress NodeToClientVersion
+  tracers =
+    NetworkConnectTracers
+      { nctMuxTracer = nullTracer
+      , nctHandshakeTracer = nullTracer
+      }
+
+client ::
+  MonadSTM m =>
+  ClientCodecs Block m ->
+  TQueue m (PostChainTx tx) ->
+  ChainCallback tx m ->
+  OuroborosApplication 'InitiatorMode LocalAddress LByteString m () Void
+client codecs queue callback =
+  OuroborosApplication $ \_connectionId _controlMessageSTM ->
+    [ localChainSyncMiniProtocol
+    , localTxSubmissionMiniProtocol
+    ]
+ where
+  -- TODO: Factor out the tracer work on Hydra.Network.Ouroboros and use it to
+  -- provide SendRecv traces for both protocols.
+  localChainSyncMiniProtocol =
+    MiniProtocol
+      { miniProtocolNum = MiniProtocolNum 5
+      , miniProtocolLimits = maximumMiniProtocolLimits
+      , miniProtocolRun = InitiatorProtocolOnly initiator
+      }
+   where
+    initiator =
+      MuxPeer
+        nullTracer
+        (cChainSyncCodec codecs)
+        (chainSyncClientPeer $ chainSyncClient callback)
+
+  localTxSubmissionMiniProtocol =
+    MiniProtocol
+      { miniProtocolNum = MiniProtocolNum 6
+      , miniProtocolLimits = maximumMiniProtocolLimits
+      , miniProtocolRun = InitiatorProtocolOnly initiator
+      }
+   where
+    initiator =
+      MuxPeer
+        nullTracer
+        (cTxSubmissionCodec codecs)
+        (localTxSubmissionClientPeer $ txSubmissionClient queue)
 
 chainSyncClient ::
-  ChainSyncClient header point tip m a
-chainSyncClient = undefined
+  forall m tx.
+  Monad m =>
+  ChainCallback tx m ->
+  ChainSyncClient Block (Point Block) (Tip Block) m ()
+chainSyncClient callback =
+  ChainSyncClient (pure clientStIdle)
+ where
+  -- FIXME: This won't work well with real client. Without acquiring any point
+  -- (i.e. agreeing on a common state / intersection with the server), the
+  -- server will start streaming blocks from the origin.
+  --
+  -- Since Hydra heads are supposedly always online, it may be sufficient to
+  -- simply negotiate the intersection at the current tip, and then, continue
+  -- following the chain from that tip. The head, or more exactly, this client,
+  -- would not be able to yield on chain events happening in the past, but only
+  -- events which occur after the hydra-node is started. For now, since our test
+  -- code is unable to illustrate that problem, I'll leave it as it is.
+  clientStIdle :: ClientStIdle Block (Point Block) (Tip Block) m ()
+  clientStIdle = SendMsgRequestNext clientStNext (pure clientStNext)
+
+  -- FIXME: rolling forward with a transaction does not necessarily mean that we
+  -- can't roll backward. Or said differently, the block / transactions yielded
+  -- by the server are not necessarily settled. Settlement only happens after a
+  -- while and we will have to carefully consider how we want to handle
+  -- rollbacks. What happen if an 'init' transaction is rolled back?
```

Comment:
This is something we already identified very early as an issue? Perhaps a topic to address with researchers or in the next engineering meeting, assuming we've made enough progress to be able to talk to a "real" cardano chain.